src/hotspot/share/memory/memoryReserver.cpp

```diff
@@ -0,0 +1,693 @@
+  sanity_check_arguments(size, alignment, page_size);
+
+  // Reserve the memory.
+
+  // There are basically three different cases that we need to handle:
+  // 1. Mapping backed by a file
+  // 2. Mapping backed by explicit large pages
+  // 3. Mapping backed by normal pages or transparent huge pages
+  // The first two have restrictions that requires the whole mapping to be
+  // committed up front. To record this the ReservedSpace is marked 'special'.
+
+  // == Case 1 ==
+  // This case is contained within the HeapReserver
+
+  // == Case 2 ==
+  if (use_explicit_large_pages(page_size)) {
+    // System can't commit large pages i.e. use transparent huge pages and
+    // the caller requested large pages. To satisfy this request we use
+    // explicit large pages and these have to be committed up front to ensure
+    // no reservations are lost.
+    do {
+      ReservedSpace reserved = reserve_memory_special(requested_address, size, alignment, page_size, executable);
+      if (reserved.is_reserved()) {
+        // Successful reservation using large pages.
+        return reserved;
+      }
+      page_size = os::page_sizes().next_smaller(page_size);
+    } while (page_size > os::vm_page_size());
+
+    // Failed to reserve explicit large pages, do proper logging.
+    log_on_large_pages_failure(requested_address, size);
+    // Now fall back to normal reservation.
+    assert(page_size == os::vm_page_size(), "inv");
+  }
+
+  // == Case 3 ==
+  return reserve_memory(requested_address, size, alignment, executable, mem_tag);
+}
+
+ReservedSpace MemoryReserver::reserve(char* requested_address,
+                                      size_t size,
+                                      size_t alignment,
+                                      size_t page_size,
+                                      MemTag mem_tag) {
+  return reserve(requested_address,
+                 size,
+                 alignment,
+                 page_size,
+                 !ExecMem,
+                 mem_tag);
+}
+
+
+ReservedSpace MemoryReserver::reserve(size_t size,
+                                      size_t alignment,
+                                      size_t page_size,
+                                      MemTag mem_tag) {
+  return reserve(nullptr /* requested_address */,
+                 size,
+                 alignment,
+                 page_size,
+                 mem_tag);
+}
+
+ReservedSpace MemoryReserver::reserve(size_t size,
+                                      MemTag mem_tag) {
+  // Want to use large pages where possible. If the size is
+  // not large page aligned the mapping will be a mix of
+  // large and normal pages.
+  size_t page_size = os::page_size_for_region_unaligned(size, 1);
+  size_t alignment = os::vm_allocation_granularity();
+
+  return reserve(size,
+                 alignment,
+                 page_size,
+                 mem_tag);
+}
+
+bool MemoryReserver::release(const ReservedSpace& reserved) {
```

Comment:
It seems like you've explicitly decided against this in the new API, is there a particular reason for that?

Reply:
My reasoning was that MemoryReserver does not have the ownership of the ReservedSpace instance, so I didn't want it to be the one clearing the members. If others feel the same way as you do, then I would prefer to add a function release_and_clear to make it clearer (:)) what's going on.

Reply:
`release_and_clear`does sound clearer :)

Reply:
My preference is to stay with the current proposal and let code that requested reserve memory decide if and when the ReservedSpace should be cleared. I think it would be an easy thing to reevaluate in another RFE if we change our minds later on.